File: 01/1.16/fast-expt.ex
defmodule FastExponentiation do
  require Integer

  # Private helper function (no @spec needed!)
  defp square(n), do: n * n

  @spec fastExpt(integer(), integer()) :: integer()
  def fastExpt(_b, 0), do: 1
  def fastExpt(b, n) when Integer.is_even(n), do: square(fastExpt(b, div(n, 2)))
  def fastExpt(b, n), do: b * fastExpt(b, n - 1)
end

File: 01/1.16/fast-expt.hs
square :: Int -> Int
square n = n * n

fastExpt :: Int -> Int -> Int
fastExpt _ 0 = 1
fastExpt b n
  | even n = square (fastExpt b (n `div` 2))
  | otherwise = b * fastExpt b (n - 1)

main :: IO ()
main = do
  print (fastExpt 2 4)
